### PR TITLE
feat: Rewrite pre-commit hook & ktlint apply to idea into Flank Scripts

### LINKED ISSUE #1423: Re-write GitHooks and Ktlint to FlankScripts
**Author the user story for this feature**

Getting Ktlint to apply to idea will be written into Flankscripts under a new section.
this also applies to Githooks.

NewCommand section name: ContributionCommand? 
WDYT?

### PR DESCRIPTION
**Pull request by Sloox**
Fixes #1423 

Tested for Windows. 

## Checklist

- [x] Documented

### CODE REVIEW

File: flank-scripts/build.gradle.kts [hunk 1]
```diff
@@ -29,7 +29,7 @@ shadowJar.apply {
     }
 }
 // <breaking change>.<feature added>.<fix/minor change>
-version = "1.2.8"
+version = "1.2.9"
```
Comment (piotradamczyk5) [inline flank-scripts/build.gradle.kts:32]: feature added, please change to `1.3.0`

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/ContributionCommand.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package flank.scripts.contribution
+
+import com.github.ajalt.clikt.core.CliktCommand
+import com.github.ajalt.clikt.core.subcommands
+import flank.scripts.contribution.githooks.GitHooksLinkCommand
+
+object ContributionCommand : CliktCommand(name = "contribution", help = "Tasks for assisting with contribution") {
+    init {
+        subcommands(GitHooksLinkCommand)
+    }
```
Comment (piotradamczyk5) [inline flank-scripts/src/main/kotlin/flank/scripts/contribution/ContributionCommand.kt:10]: please update `README.md`

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/githooks/GitHooksLinkCommand.kt [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+import flank.scripts.utils.runCommand
+
+object GitHooksLinkCommand : CliktCommand(
+    name = "linkGitHooks",
+    help = "Creates a link for pre-commit hook for automatic linting"
+) {
+    override fun run() {
+        logLn("Linking Githooks.")
+        linkGitHooks()
+    }
```
Comment (piotradamczyk5) [inline flank-scripts/src/main/kotlin/flank/scripts/contribution/githooks/GitHooksLinkCommand.kt:14]: please update `README.md`

File: flank-scripts/src/test/kotlin/flank/common/FilesTest.kt [hunk 4]
```diff
@@ -10,6 +11,7 @@ internal class FilesTest {
 
     @Test
     fun `Should create symbolic file at desired location`() {
+        Assume.assumeFalse(isWindows)
```
Comment (piotradamczyk5) [inline flank-scripts/src/test/kotlin/flank/common/FilesTest.kt:14]: does it fail on windows? I assume it should work cause under the hood it uses java api

Comment (Sloox) [reply flank-scripts/src/test/kotlin/flank/common/FilesTest.kt:14]: It does. https://github.com/Flank/flank/issues/1434

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/ContributionCommand.kt (hunk 2 above)
Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ContributionCommand.kt:14]: 👍

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/githooks/GitHooksLinkCommand.kt (hunk 3 above)
Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/githooks/GitHooksLinkCommand.kt:14]: 👍

File: flank-scripts/build.gradle.kts (hunk 1 above)
Comment (Sloox) [reply flank-scripts/build.gradle.kts]: done!

File: flank-scripts/src/test/kotlin/flank/common/FilesTest.kt (hunk 4 above)
Comment (piotradamczyk5) [reply flank-scripts/src/test/kotlin/flank/common/FilesTest.kt:14]: oh I see, I tested it on CI, so it does not have problem with permissions

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+import java.io.File
+
+object IdeaKtlintCodeStyleCommand : CliktCommand(
+    name = "applyKtlintToIdea",
+    help = "Applies Ktlint to this idea project forcefully"
+) {
+    override fun run(): Unit = runBlocking {
+        logLn("Applying Ktlint code style to this idea project")
+        logLn("Retrieving Ktlint...")
+        when (retrieveKtlintCurl()) {
```
Comment (pawelpasterz) [inline flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:19]: Why 2 ways of downloading ktlint? Can't we just use `retrieveKtlintResolve` ?

Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:19]: Windows doesnt always have a bash or mingw backend therefore curl isnt always avaialble. The second is a backup 👍

Comment (pawelpasterz) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:19]: Got it. How about using `retrieveKtlintResolve` only? A bit less logic and it's working regardless of OS

Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:20]: @pawelpasterz i will test it :) I was looking for a nice progress bar. Doing it with Fuel creates way more logic and for me specifically i have curl installed on windows so it generally looks good. Its for edge cases.

Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:20]: have bigger issues with pathing at the moment than anything else.

Comment (piotradamczyk5) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:20]: @Sloox  if it is possible please use the non-curl version for all OS.

Comment (bootstraponline) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:20]: +1 to investing in the Kotlin logic that works for all OS instead of depending on curl.

Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:20]: Ill remove the curl command

File: flank-scripts/build.gradle.kts [hunk 6]
```diff
@@ -28,7 +28,7 @@ shadowJar.apply {
     }
 }
 // <breaking change>.<feature added>.<fix/minor change>
-version = "1.3.1"
+version = "1.3.2"
```
Comment (piotradamczyk5) [inline flank-scripts/build.gradle.kts:31]: feature added > `1.4.0`

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        logLn("Cleanup Ktlint leftover files...")
+        (getKtlintFilePath()).deleteFile()
+    }
+
+    private suspend fun retrieveKtlintResolveCommand() = "https://github.com/pinterest/ktlint/releases/download/0.40.0/ktlint"
+        .httpDownload().fileDestination { _, _ ->
+            Files.createFile(Paths.get(getKtlintFilePath())).toFile().also {
+                logLn("Ktlint written to: ${it.absolutePath}")
+            }
+        }.awaitUnit()
```
Comment (piotradamczyk5) [inline flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:44]: we have similiar function in `:common` module

File: flank-scripts/build.gradle.kts (hunk 6 above)
Comment (piotradamczyk5) [inline flank-scripts/build.gradle.kts:31]: feature added -> `1.4.0`

**Review by piotradamczyk5 — CHANGES_REQUESTED** [review]

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt (hunk 7 above)
Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:44]: @piotradamczyk5 can you show me where?

File: flank-scripts/build.gradle.kts (hunk 6 above)
Comment (Sloox) [reply flank-scripts/build.gradle.kts:31]: hmmm i did. From 1.2 -.13 :P then after each PR it requires a minor change.

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt (hunk 7 above)
Comment (piotradamczyk5) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:44]: https://github.com/Flank/flank/blob/7059c10ccf748cef4c275473562a164d1783692c/common/src/main/kotlin/flank.common/Files.kt#L31

Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:44]: Ahh I misunderstood what you meant then.

**Comment by adamfilipow92** [comment]
I try to test how pre-commit hook works. 
1. Add unused import to ```ValidateIosArgs.kt```
2. Commit changes
3. Pre-commit hooks works 👍 
4. I can see in committed changes this unused import
5. In uncommitted changes I have ```ValidateIosArgs.kt``` with removed import

Maybe it's should commit files changed by ktlint? I'm not sure how it works before. WDYT?

File: flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt (hunk 7 above)
Comment (Sloox) [reply flank-scripts/src/main/kotlin/flank/scripts/contribution/ideaktlint/IdeaKtlintCodeStyleCommand.kt:44]: @piotradamczyk5 it didn't work for the use case so I made a new function.

**Comment by Sloox** [comment]
> I try to test how pre-commit hook works.
> 
> 1. Add unused import to `ValidateIosArgs.kt`
> 2. Commit changes
> 3. Pre-commit hooks works 👍
> 4. I can see in committed changes this unused import
> 5. In uncommitted changes I have `ValidateIosArgs.kt` with removed import
> 
> Maybe it's should commit files changed by ktlint? I'm not sure how it works before. WDYT?

@adamfilipow92 can you explain a bit more what you mean i am confused.

**Review by piotradamczyk5 — APPROVED** [review]

**Review by pawelpasterz — APPROVED** [review]
